From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 23 Mar 2025 15:16:05 +0800
Subject: [PATCH] Pufferfish Reduce chunk loading & lookups


diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index 5ae08be75ca01924fc78bdd8d6bb6747ddc21aea..af2f2674bec98e6958d9683a601d99f0fb030f7f 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -300,11 +300,17 @@ public class EnderMan extends Monster implements NeutralMob {
     private boolean teleport(double x, double y, double z) {
         BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos(x, y, z);
 
-        while (mutableBlockPos.getY() > this.level().getMinY() && !this.level().getBlockState(mutableBlockPos).blocksMotion()) {
+        // Pufferfish start - single chunk lookup
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(mutableBlockPos);
+        if (chunk == null) {
+            return false;
+        }
+        // Pufferfish end
+        while (mutableBlockPos.getY() > this.level().getMinY() && !chunk.getBlockState(mutableBlockPos).blocksMotion()) { // Pufferfish
             mutableBlockPos.move(Direction.DOWN);
         }
 
-        BlockState blockState = this.level().getBlockState(mutableBlockPos);
+        BlockState blockState = chunk.getBlockState(mutableBlockPos); // Pufferfish
         boolean flag = blockState.blocksMotion();
         boolean isWater = blockState.getFluidState().is(FluidTags.WATER);
         if (flag && !isWater) {
