From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Tue, 21 Apr 2026 23:52:18 +0800
Subject: [PATCH] Pufferfish Optimize suffocation


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index f1db892a9d6a27889e983eae3e4c029b8c44315c..9c063bf4b893ed5fd6fbbf99847bfafe2636794d 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -454,7 +454,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         profiler.push("livingEntityBaseTick");
         if (this.isAlive() && this.level() instanceof ServerLevel level) {
             boolean isPlayer = this instanceof Player;
-            if (this.isInWall()) {
+            if (this.shouldCheckForSuffocation() && this.isInWall()) { // Pufferfish - optimize suffocation
                 this.hurtServer(level, this.damageSources().inWall(), 1.0F);
             } else if (isPlayer && !level.getWorldBorder().isWithinBounds(this.getBoundingBox())) {
                 double dist = level.getWorldBorder().getDistanceToBorder(this) + level.getWorldBorder().getSafeZone();
@@ -1442,6 +1442,22 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return this.getHealth() <= 0.0F;
     }
 
+    // Pufferfish start - optimize suffocation
+    public boolean couldPossiblyBeHurt(float amount) {
+        if ((float) this.invulnerableTime > (float) this.invulnerableDuration / 2.0F && amount <= this.lastHurt) {
+            return false;
+        }
+        return true;
+    }
+    public boolean shouldCheckForSuffocation() {
+        return !dev.bacteriawa.mint.config.modules.optimizations.SuffocationOptimizationConfig.enabled || (this.tickCount % 10 == 0 && this.couldPossibilyBeHurt(1.0F));
+    }
+
+    public final boolean couldPossibilyBeHurt(float amount) { // Mint - typo fix/match patch
+        return this.couldPossiblyBeHurt(amount);
+    }
+    // Pufferfish end
+
     @Override
     public boolean hurtServer(final ServerLevel level, final DamageSource source, float damage) {
         if (this.isInvulnerableTo(level, source)) {
diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 419c29c003b3e690a89a14c0a9b3eb2ac3500175..f64acc45f389864092f25f19a3fe2513508cfe96 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -133,6 +133,13 @@ public class WitherBoss extends Monster implements RangedAttackMob {
         }
     }
 
+    // Pufferfish start - optimize suffocation
+    @Override
+    public boolean shouldCheckForSuffocation() {
+        return true;
+    }
+    // Pufferfish end
+
     @Override
     public void setCustomName(final @Nullable Component name) {
         super.setCustomName(name);
