From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 4 May 2026 14:46:02 +0800
Subject: [PATCH] Do not search the block out of current region


diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 832e04b97ffae60127537c0a8eb448af6602da1e..dcacb2135e2e549b8929ca7e173267fc24c537e5 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -427,6 +427,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
                 if (xd != 0 || zd != 0 || allowBedrock) {
                     for (int y = level.getMaxY(); y > (tallest == null ? level.getMinY() : tallest.getY()); y--) {
                         BlockPos pos = new BlockPos(around.getX() + xd, y, around.getZ() + zd);
+                        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor((Level) level, pos)) continue; // Mint - Do not search the block out of current region as we might be at the edge of the tickregion
                         BlockState state = level.getBlockState(pos);
                         if (state.isCollisionShapeFullBlock(level, pos) && (allowBedrock || !state.is(Blocks.BEDROCK))) {
                             tallest = pos;
