From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Tue, 1 Apr 2025 00:52:57 +0800
Subject: [PATCH] Do not search the block out of current region


diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 12b9a6c16501cb0c01983c44bb247071ce675274..064de41baa97af9e0eeeef86f786e6ee21a7f8de 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -422,6 +422,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
                 if (i != 0 || i1 != 0 || allowBedrock) {
                     for (int y = level.getMaxY(); y > (blockPos == null ? level.getMinY() : blockPos.getY()); y--) {
                         BlockPos blockPos1 = new BlockPos(pos.getX() + i, y, pos.getZ() + i1);
+                        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor((Level) level, blockPos1)) continue; // Mint - Do not search the block out of current region as we might be at the edge of the tickregion
                         BlockState blockState = level.getBlockState(blockPos1);
                         if (blockState.isCollisionShapeFullBlock(level, blockPos1) && (allowBedrock || !blockState.is(Blocks.BEDROCK))) {
                             blockPos = blockPos1;
