From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 26 Apr 2026 21:20:14 +0800
Subject: [PATCH] Gale Skip entity move if movement is zero


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index f29ee350a02055f93a2ad8d2f3433cfb60f556ab..e7669272a25afdc06888bc6e87801a2896021e34 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1148,6 +1148,8 @@ public abstract class Entity
     private double moveStartZ;
     // Paper end - detailed watchdog information
 
+    private boolean boundingBoxChanged = false; // Gale - VMP - skip entity move if movement is zero
+
     public void move(final MoverType moverType, Vec3 delta) {
         final Vec3 originalMovement = delta; // Paper - Expose pre-collision velocity
         // Paper start - detailed watchdog information
@@ -1160,6 +1162,12 @@ public abstract class Entity
         }
         try {
         // Paper end - detailed watchdog information
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.boundingBoxChanged && delta.equals(Vec3.ZERO)) {
+            return;
+        }
+        this.boundingBoxChanged = false;
+        // Gale end - VMP - skip entity move if movement is zero
         if (this.noPhysics) {
             this.setPos(this.getX() + delta.x, this.getY() + delta.y, this.getZ() + delta.z);
             this.horizontalCollision = false;
@@ -5322,6 +5330,11 @@ public abstract class Entity
     }
 
     public final void setBoundingBox(final AABB bb) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.bb.equals(bb)) {
+            this.boundingBoxChanged = true;
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
                 minY = bb.minY,
