From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 26 Apr 2026 21:33:57 +0800
Subject: [PATCH] Leaf Paper PR: Prevent zombie reinforcements loading chunks


diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index 0008aa0ca02fc4ea67b795b3f0a0e9656233ce89..38c2f7c04c4e464947b5add19b8e292314613c24 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -342,6 +342,11 @@ public class Zombie extends Monster {
                     int yt = y + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     int zt = z + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     BlockPos spawnPos = new BlockPos(xt, yt, zt);
+                    // Paper start - Prevent reinforcement checks from loading chunks
+                    if (level.getChunkIfLoadedImmediately(spawnPos.getX() >> 4, spawnPos.getZ() >> 4) == null) {
+                        continue;
+                    }
+                    // Paper end - Prevent reinforcement checks from loading chunks
                     if (SpawnPlacements.isSpawnPositionOk(type, level, spawnPos)
                         && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.REINFORCEMENT, spawnPos, level.getRandom())) {
                         reinforcement.setPos(xt, yt, zt);
