From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 12 Apr 2025 18:28:16 +0800
Subject: [PATCH] Add config for server mod name


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 73346264aa1683b376a42d56db384448c3795cf7..6b055e7d2a6e837e7895417ba4f048bee5040d79 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2099,7 +2099,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return dev.bacteriawa.mint.config.modules.misc.ServerBrandNameConfig.serverModName; // Mint
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
