From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 12 Apr 2025 18:28:16 +0800
Subject: [PATCH] Add config for server mod name


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index d50ac3fad39dc965b8d5c65e904dded492ec6051..3e2114aaa624bc37d471f88b11286127c9aa5f2e 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2101,7 +2101,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return dev.bacteriawa.mint.config.modules.misc.ServerBrandNameConfig.serverModName; // Mint
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
