From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Fri, 19 Dec 2025 23:25:11 +0800
Subject: [PATCH] Add config for username check


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 4fe9bfcf75db1b5401776711a5cccde9e1feb743..b318c71bbf73f0370e526aaa0c14e6e59bcf245a 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -159,7 +159,8 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
     public void handleHello(ServerboundHelloPacket packet) {
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.HELLO, "Unexpected hello packet");
         // Paper start - Validate usernames
-        if (io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode()
+        if (dev.bacteriawa.mint.config.modules.misc.UsernameCheckConfig.enabled // Mint - Add config for username check
+            && io.papermc.paper.configuration.GlobalConfiguration.get().proxies.isProxyOnlineMode()
             && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.performUsernameValidation
             && !this.iKnowThisMayNotBeTheBestIdeaButPleaseDisableUsernameValidation) {
             Validate.validState(StringUtil.isReasonablePlayerName(packet.name()), "Invalid characters in username");
diff --git a/net/minecraft/util/StringUtil.java b/net/minecraft/util/StringUtil.java
index bd57f9a4224fc236cccecfd13b03f27d20ed15e9..4a65bd62d1f97bd48b87356a2b305179455f8b44 100644
--- a/net/minecraft/util/StringUtil.java
+++ b/net/minecraft/util/StringUtil.java
@@ -62,8 +62,14 @@ public class StringUtil {
     public static boolean isAllowedChatCharacter(int codePoint) {
         return codePoint != 167 && codePoint >= 32 && codePoint != 127;
     }
+    // Mint start - Add config for username checks
+    public static boolean isValidPlayerName(String username){
+        return isValidPlayerName(username, !dev.bacteriawa.mint.config.modules.misc.UsernameCheckConfig.enabled);
+    }
+    // Mint end
 
-    public static boolean isValidPlayerName(String playerName) {
+    public static boolean isValidPlayerName(String playerName, boolean byPass) { // Mint - Add config for username checks
+        if (byPass) return playerName.length() <= 16; // Mint - Add config for username checks
         return playerName.length() <= 16 && playerName.chars().filter(i -> i <= 32 || i >= 127).findAny().isEmpty();
     }
 
