From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 2 May 2025 12:03:14 +0800
Subject: [PATCH] Luminol Fix incorrectly synced data after player
 teleportation


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 4e3c545f6121d3846d81d776c29edc42d6204f5c..ab26728e5a29a2d7e58860e57778cf60c9e1767d 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1834,14 +1834,20 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             this.connection.send(new ClientboundChangeDifficultyPacket(
                 worlddata.getDifficulty(), worlddata.isDifficultyLocked()
             ));
-            this.connection.send(new ClientboundSetExperiencePacket(
+            /*this.connection.send(new ClientboundSetExperiencePacket( // Luminol start - teleportAsync tweaks, will send in next tick, see the changes below
                 this.experienceProgress, this.totalExperience, this.experienceLevel
-            ));
+            ));*/ // Luminol end
 
             playerlist.sendActivePlayerEffects(this);
             playerlist.sendLevelInfo(this, destination);
             playerlist.sendPlayerPermissionLevel(this);
 
+            // Luminol start - teleportAsync tweaks
+            this.lastSentExp = -1;
+            this.lastSentHealth = -1.0F;
+            this.lastSentFood = -1;
+            // Luminol end
+
             // regular world add logic
             this.unsetRemoved();
             destination.addDuringTeleport(this);
