From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 25 May 2025 03:15:22 +0800
Subject: [PATCH] Leaf Remove stream in villagers


diff --git a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
index 623dd33a85cb67ecb835de18e9aac29f3394ce28..343d5f0666165c773c893261b2b6a6045b2c85e3 100644
--- a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -71,9 +71,19 @@ public class TradeWithVillager extends Behavior<Villager> {
     }
 
     private static Set<Item> figureOutWhatIAmWillingToTrade(Villager villager, Villager other) {
-        ImmutableSet<Item> set = other.getVillagerData().profession().value().requestedItems();
-        ImmutableSet<Item> set1 = villager.getVillagerData().profession().value().requestedItems();
-        return set.stream().filter(item -> !set1.contains(item)).collect(Collectors.toSet());
+        // Leaf start - Remove stream in villagers
+        ImmutableSet<Item> otherItems = other.getVillagerData().profession().value().requestedItems();
+        ImmutableSet<Item> villagerItems = villager.getVillagerData().profession().value().requestedItems();
+        Set<Item> result = new java.util.HashSet<>();
+
+        for (Item item : otherItems) {
+            if (!villagerItems.contains(item)) {
+                result.add(item);
+            }
+        }
+
+        return result;
+        // Leaf end - Remove stream in villagers
     }
 
     private static void throwHalfStack(Villager villager, Set<Item> stack, LivingEntity entity) {
diff --git a/net/minecraft/world/entity/npc/Villager.java b/net/minecraft/world/entity/npc/Villager.java
index a8fa795a12e978230b01c2022176be4c57eba2f4..73c64a15cdad94ae7809ce1a865e202d751323c6 100644
--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -863,7 +863,17 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Leaf start - Remove stream in villagers
+        int sum = 0;
+
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            Item item = entry.getKey();
+            int points = entry.getValue();
+            sum += inventory.countItem(item) * points;
+        }
+
+        return sum;
+        // Leaf end - Remove stream in villagers
     }
 
     public boolean hasFarmSeeds() {
