From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 20 Dec 2025 01:06:54 +0800
Subject: [PATCH] Add config for command block command execution


diff --git a/net/minecraft/world/level/BaseCommandBlock.java b/net/minecraft/world/level/BaseCommandBlock.java
index 938d3bf93f7f48c07cfdc5de2abf6a635b1ae650..1a8fbe53147770e32b7b534d67a4b6d506c8456d 100644
--- a/net/minecraft/world/level/BaseCommandBlock.java
+++ b/net/minecraft/world/level/BaseCommandBlock.java
@@ -91,7 +91,7 @@ public abstract class BaseCommandBlock {
     }
 
     public boolean performCommand(ServerLevel level) {
-        if (true) return false; // Folia - region threading
+        if (!dev.bacteriawa.mint.config.modules.experiment.CommandBlockConfig.enabled) return false; // Folia - region threading // Mint
         if (level.getGameTime() == this.lastExecution) {
             return false;
         } else if ("Searge".equalsIgnoreCase(this.command)) {
