From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 25 Jun 2025 01:14:50 +0800
Subject: [PATCH] Leaf Optimize BlockEntityType#isValid


diff --git a/net/minecraft/world/level/block/entity/BlockEntityType.java b/net/minecraft/world/level/block/entity/BlockEntityType.java
index 7b8631abf8adcdb889d412b30eca10faf044032a..a2c01595053e6e7b2d6fdc6c39501ff8252e3168 100644
--- a/net/minecraft/world/level/block/entity/BlockEntityType.java
+++ b/net/minecraft/world/level/block/entity/BlockEntityType.java
@@ -260,6 +260,14 @@ public class BlockEntityType<T extends BlockEntity> {
         // Folia end - profiler
         this.factory = factory;
         this.validBlocks = validBlocks;
+        // Leaf start - Optimize BlockEntityType#isValid
+        for (Block validBlock : validBlocks) {
+            if (validBlock.blockEntityType != null) {
+                throw new IllegalStateException("Duplicate block entity type");
+            }
+            validBlock.blockEntityType = this;
+        }
+        // Leaf end - Optimize BlockEntityType#isValid
     }
 
     public T create(BlockPos pos, BlockState state) {
@@ -267,7 +275,7 @@ public class BlockEntityType<T extends BlockEntity> {
     }
 
     public boolean isValid(BlockState state) {
-        return this.validBlocks.contains(state.getBlock());
+        return state.getBlock().blockEntityType == this; // Leaf - Optimize BlockEntityType#isValid - remove hash lookup
     }
 
     @Deprecated
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 7619061ff78483ea8feb166a0a0ea20f2576006e..ce7ceff24fcec2c3d2b20f0cd1705a03e4b8123f 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -101,6 +101,7 @@ public abstract class BlockBehaviour implements FeatureElement {
     protected final BlockBehaviour.Properties properties;
     protected final Optional<ResourceKey<LootTable>> drops;
     protected final String descriptionId;
+    @Nullable public net.minecraft.world.level.block.entity.BlockEntityType blockEntityType = null; // Leaf - Optimize BlockEntityType#isValid
 
     public BlockBehaviour(BlockBehaviour.Properties properties) {
         this.hasCollision = properties.hasCollision;
