From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Frish2021 <1573880184@qq.com>
Date: Wed, 9 Jul 2025 19:51:53 +0800
Subject: [PATCH] Fix Commands class can`t used config system.


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index fd3553bdc1c3cdbf6aa3dc00e0a4987f8eaa4fb8..a6d11446fe5791254047b88076371f801b9d4435 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -16,6 +16,8 @@ import java.nio.file.Paths;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+
+import dev.bacteriawa.mint.config.MintConfig;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.OptionSpec;
@@ -109,6 +111,7 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
+            MintConfig.loadConfig(); // Mint
             io.papermc.paper.plugin.PluginInitializerManager.load(optionSet); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 3fb73cf62f51401ea0d3241e0f9fbb6b30cf7f62..93b47367baf31c17400d56651849427e7f949c82 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -20,7 +20,13 @@ import java.util.Locale;
 import java.util.Optional;
 import javax.annotation.Nullable;
 
+import dev.bacteriawa.mint.commands.MemoryBarCommand;
+import dev.bacteriawa.mint.commands.RegionBarCommand;
+import dev.bacteriawa.mint.commands.TpsBarCommand;
 import dev.bacteriawa.mint.config.MintConfig;
+import dev.bacteriawa.mint.functions.GlobalServerMemoryBar;
+import dev.bacteriawa.mint.functions.GlobalServerRegionBar;
+import dev.bacteriawa.mint.functions.GlobalServerTpsBar;
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
 import net.minecraft.SharedConstants;
@@ -59,6 +65,7 @@ import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.entity.SkullBlockEntity;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.bukkit.Bukkit;
 import org.slf4j.Logger;
 
 public class DedicatedServer extends MinecraftServer implements ServerInterface {
@@ -173,8 +180,29 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
-        MintConfig.loadConfig(); // Mint
         MintConfig.setup(); // Mint
+
+        if (dev.bacteriawa.mint.config.modules.misc.TpsBarConfig.tpsbarEnabled){
+            GlobalServerTpsBar.init();
+            Bukkit.getCommandMap().register("tpsbar","mint",new TpsBarCommand());
+        }else{
+            GlobalServerTpsBar.cancelBarUpdateTask();
+        }
+
+        if (dev.bacteriawa.mint.config.modules.misc.MembarConfig.memoryBarEnabled){
+            GlobalServerMemoryBar.init();
+            Bukkit.getCommandMap().register("membar","mint",new MemoryBarCommand());
+        }else{
+            GlobalServerMemoryBar.cancelBarUpdateTask();
+        }
+
+        if (dev.bacteriawa.mint.config.modules.misc.RegionBarConfig.regionbarEnabled) {
+            GlobalServerRegionBar.init();
+            Bukkit.getCommandMap().register("regionbar", "mint", new RegionBarCommand());
+        } else {
+            GlobalServerRegionBar.cancelBarUpdateTask();
+        }
+
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
