From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 27 Jul 2025 17:06:17 +0800
Subject: [PATCH] Add Networkbar


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index 1626b932a5e46ee77465ef08c2f4430c2e2753c1..53187901756b35b45495c2556c942f6ffd280216 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -66,6 +66,8 @@ import org.slf4j.Logger;
 import org.slf4j.Marker;
 import org.slf4j.MarkerFactory;
 
+import static dev.bacteriawa.mint.functions.GlobalServerNetworkBar.updatePlayerNetworkData;
+
 public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
     private static final float AVERAGE_PACKETS_SMOOTHING = 0.75F;
     private static final Logger LOGGER = LogUtils.getLogger();
@@ -328,6 +330,12 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
                 // Paper end - packet limiter
                 if (packetListener.shouldHandleMessage(packet)) {
                     try {
+                        // Mint start - Update player network data for received packets
+                        net.minecraft.server.level.ServerPlayer player = this.getPlayer();
+                        if (player != null) {
+                            updatePlayerNetworkData(player.getBukkitEntity(), packet.toString().length(), 0, 1, 0);
+                        }
+                        // Mint end - Update player network data for received packets
                         genericsFtw(packet, packetListener);
                     } catch (RunningOnDifferentThreadException var5) {
                     } catch (io.papermc.paper.util.ServerStopRejectedExecutionException ignored) { // Paper - do not prematurely disconnect players on stop
@@ -523,6 +531,11 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
             return;
         }
         try {
+        // Mint start - Update player network data for sent packets
+        if (player != null) {
+            updatePlayerNetworkData(player.getBukkitEntity(), 0, packet.toString().length(), 0, 1);
+        }
+        // Mint end - Update player network data for sent packets
         final ChannelFuture channelFuture;
         // Paper end - Optimize network
         if (channelFutureListener != null) {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 8da291c8171ee453cba13fbe8fde50fbc4c4b40d..4003c9f71d4a8c32a02ae6eb3ef7b59ca463470a 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -732,6 +732,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         dev.bacteriawa.mint.functions.GlobalServerTpsBar.cancelBarUpdateTask(); //Luminol - Tpsbar
         dev.bacteriawa.mint.functions.GlobalServerMemoryBar.cancelBarUpdateTask(); //Luminol - Memory bar
         dev.bacteriawa.mint.functions.GlobalServerRegionBar.cancelBarUpdateTask(); //Luminol - Region bar
+        dev.bacteriawa.mint.functions.GlobalServerNetworkBar.cancelBarUpdateTask(); // Mint - Network bar
         super.stopServer();
         //Util.shutdownExecutors(); // Paper - Improved watchdog support; moved into super
         SkullBlockEntity.clear();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 886299f6a69e57286b6acbee416b552163f2c5b2..03be04329f392b7d8ff002bb7bc6147bc32c11f1 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -432,6 +432,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public volatile boolean isTpsBarVisible = false; //Luminol - Tps bar
     public volatile boolean isMemBarVisible = false; //Luminol - Memory bar
     public volatile boolean isRegionBarVisible = false; //Luminol - Region bar
+    public volatile boolean isNetworkBarVisible = false; // Mint - Network bar
     // Paper start - rewrite chunk system
     private ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.PlayerChunkLoaderData chunkLoader;
     private final ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder viewDistanceHolder = new ca.spottedleaf.moonrise.patches.chunk_system.player.RegionizedPlayerChunkLoader.ViewDistanceHolder();
