From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 4 May 2026 22:53:06 +0800
Subject: [PATCH] PufferFish Dynamic Activation Brain


diff --git a/io/papermc/paper/entity/activation/ActivationRange.java b/io/papermc/paper/entity/activation/ActivationRange.java
index 7e5571f986bb739b6b39134507afa1b3dc0a29c8..97ac1e01b20196669573f22f8c10574fe8f30bb5 100644
--- a/io/papermc/paper/entity/activation/ActivationRange.java
+++ b/io/papermc/paper/entity/activation/ActivationRange.java
@@ -280,6 +280,15 @@ public final class ActivationRange {
             if (entity instanceof final Mob mob && mob.targetSelector.hasTasks()) {
                 return 0;
             }
+
+            // PufferFish DAB start
+            if (dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.enabled && entity.getType().dabEnabled) {
+                if (inactiveFor > dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.activationThreshold) {
+                    return 20;
+                }
+            }
+            // PufferFish DAB end
+
             if (entity instanceof final Pillager pillager) {
                 // TODO:?
             }
@@ -315,6 +324,16 @@ public final class ActivationRange {
         boolean isActive = entity.activatedTick >= io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick(); // Folia - threaded regions
         entity.isTemporarilyActive = false;
 
+        // PufferFish DAB start
+        if (!isActive && entity.getType().dabEnabled) {
+            if (entity.activatedPriorityReset) {
+                entity.activatedPriorityReset = false;
+                entity.activatedPriority = dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.maximumActivationPrio;
+            }
+            isActive = (io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + entity.getId()) % entity.activatedPriority == 0;
+        }
+        // PufferFish DAB end
+
         // Should this entity tick?
         if (!isActive) {
             if ((io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() - entity.activatedTick - 1) % 20 == 0) { // Folia - threaded regions
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 514afede981e6de54d28c1a5d89c36e11ef36a4b..c6cdd5f21ccde72f923b6fcbe4eb24a984b65449 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -937,9 +937,12 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                                         entity.stopRiding();
                                     }
 
-                                    profiler.push("tick");
-                                    this.guardEntityTick(this::tickNonPassenger, entity);
-                                    profiler.pop();
+                                profiler.push("tick");
+                                if (entity.activatedPriority > dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.maximumActivationPrio) { // PufferFish DAB
+                                    entity.activatedPriority = dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.maximumActivationPrio;
+                                }
+                                this.guardEntityTick(this::tickNonPassenger, entity);
+                                profiler.pop();
                                 }
                             }
                         }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 33b9ab75fc1d6cd42fe42514e455e78eec07f2c4..5857eb3d5850aa6059c6410f1aaf12ec57a956f4 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -390,6 +390,8 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public boolean activatedPriorityReset = false; // PufferFish DAB
+    public int activatedPriority = dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.maximumActivationPrio; // PufferFish DAB
 
     public void inactiveTick() {
     }
diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index cb0a52dc3795e1cf54069bcbe82abf7877a658b8..ce3d785c92819b9c8196ff84ed9277c4d9aac2d3 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1225,6 +1225,7 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
     private final float spawnDimensionsScale;
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
+    public boolean dabEnabled = false; // PufferFish DAB
 
     private static <T extends Entity> EntityType<T> register(final ResourceKey<EntityType<?>> id, final EntityType.Builder<T> builder) {
         return Registry.register(BuiltInRegistries.ENTITY_TYPE, id, builder.build(id));
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 6a179d9268b4a82eb7721b2cb6081eb63b4ce455..22af2a2a529f2815d548cd3a38d28a8bd023300b 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -216,10 +216,10 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
     public void inactiveTick() {
         super.inactiveTick();
         if (!this.aware) return; // Paper - Do not tick AI for inactive unaware mobs
-        if (this.goalSelector.inactiveTick()) {
+        if (this.goalSelector.inactiveTick(this.activatedPriority, true)) { // PufferFish DAB
             this.goalSelector.tick();
         }
-        if (this.targetSelector.inactiveTick()) {
+        if (this.targetSelector.inactiveTick(this.activatedPriority, true)) { // PufferFish DAB
             this.targetSelector.tick();
         }
     }
@@ -829,16 +829,20 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
         profiler.pop();
         if (i % 2 != 0 && this.tickCount > 1) {
             profiler.push("targetSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // PufferFish DAB
             this.targetSelector.tickRunningGoals(false);
             profiler.pop();
             profiler.push("goalSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // PufferFish DAB
             this.goalSelector.tickRunningGoals(false);
             profiler.pop();
         } else {
             profiler.push("targetSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // PufferFish DAB
             this.targetSelector.tick();
             profiler.pop();
             profiler.push("goalSelector");
+            if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // PufferFish DAB
             this.goalSelector.tick();
             profiler.pop();
         }
diff --git a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
index bb037ea27688a06bb0d8141ef25d2543d34aba80..b67bbcf77e7449958c6d7eb67e26997e29da1609 100644
--- a/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerPanicTrigger.java
@@ -21,6 +21,7 @@ public class VillagerPanicTrigger extends Behavior<Villager> {
     @Override
     protected void start(final ServerLevel level, final Villager body, final long timestamp) {
         if (isHurt(body) || hasHostile(body)) {
+            body.activatedPriority = dev.bacteriawa.mint.config.modules.optimizations.DynamicActivationBrainConfig.maximumActivationPrio; // PufferFish DAB
             Brain<?> brain = body.getBrain();
             if (!brain.isActive(Activity.PANIC)) {
                 brain.eraseMemory(MemoryModuleType.PATH);
diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 75d8953845b2e9acb62b19a859d0e39fd63533eb..4574cef8c2b6854993f8877916ca18f109813247 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -81,6 +81,10 @@ public class GoalSelector {
         return true;
     }
 
+    public boolean inactiveTick(int activationPrio, boolean inactive) {
+        return (io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + (inactive ? 0 : 1)) % activationPrio == 0;
+    }
+
     public void tick() {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("goalCleanup");
diff --git a/net/minecraft/world/entity/animal/allay/Allay.java b/net/minecraft/world/entity/animal/allay/Allay.java
index 25b6e4904f0dff6159f3fb99e4226a397ef6a72b..4ea163b7a80d4bb954b7825e5e8041053782138d 100644
--- a/net/minecraft/world/entity/animal/allay/Allay.java
+++ b/net/minecraft/world/entity/animal/allay/Allay.java
@@ -209,7 +209,7 @@ public class Allay extends PathfinderMob implements InventoryCarrier, VibrationS
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("allayBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("allayActivityUpdate");
         AllayAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/armadillo/Armadillo.java b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
index 03ec05a860781f85e58055d6f58893963781bd97..906e24357ff85f75f52dd33f09155161436ee0ec 100644
--- a/net/minecraft/world/entity/animal/armadillo/Armadillo.java
+++ b/net/minecraft/world/entity/animal/armadillo/Armadillo.java
@@ -139,7 +139,7 @@ public class Armadillo extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("armadilloBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("armadilloActivityUpdate");
         ArmadilloAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/axolotl/Axolotl.java b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
index bcbe1d44871124e0b9b453cb7b7c453fa0144bc6..3e8da24ade3d89067117f5bce982c93b85449abb 100644
--- a/net/minecraft/world/entity/animal/axolotl/Axolotl.java
+++ b/net/minecraft/world/entity/animal/axolotl/Axolotl.java
@@ -366,7 +366,7 @@ public class Axolotl extends Animal implements Bucketable {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("axolotlBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("axolotlActivityUpdate");
         AxolotlAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/camel/Camel.java b/net/minecraft/world/entity/animal/camel/Camel.java
index 1aa18543428578ff773a32c24f33b843b10e8e27..91a92ea322e47a3bc32b427c7c00f571d551b9d1 100644
--- a/net/minecraft/world/entity/animal/camel/Camel.java
+++ b/net/minecraft/world/entity/animal/camel/Camel.java
@@ -166,8 +166,7 @@ public class Camel extends AbstractHorse {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("camelBrain");
-        Brain<Camel> brain = this.getBrain();
-        brain.tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("camelActivityUpdate");
         CamelAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/frog/Frog.java b/net/minecraft/world/entity/animal/frog/Frog.java
index 28b4e183efd851e6d2bb5bb16904817d4c40273b..4b5445e37162c455e6555841a2a2026842331b56 100644
--- a/net/minecraft/world/entity/animal/frog/Frog.java
+++ b/net/minecraft/world/entity/animal/frog/Frog.java
@@ -176,7 +176,7 @@ public class Frog extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("frogBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("frogActivityUpdate");
         FrogAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/frog/Tadpole.java b/net/minecraft/world/entity/animal/frog/Tadpole.java
index 3d4bfeae2fddb9ace26e7dea8870704d8309862b..07795417299c41c93310e75b9a3d1843dfabd7f9 100644
--- a/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -85,7 +85,7 @@ public class Tadpole extends AbstractFish {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("tadpoleBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("tadpoleActivityUpdate");
         TadpoleAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/goat/Goat.java b/net/minecraft/world/entity/animal/goat/Goat.java
index 22123232e0f5fe60cc9629cde7cf56bd464932d0..d4baaf37b0807949cc1df7f9726cf4c679cb8097 100644
--- a/net/minecraft/world/entity/animal/goat/Goat.java
+++ b/net/minecraft/world/entity/animal/goat/Goat.java
@@ -170,7 +170,7 @@ public class Goat extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("goatBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         profiler.push("goatActivityUpdate");
         GoatAi.updateActivity(this);
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index c947f26f57de71ac0eafc0f56a96fec12ce5dba5..688b5bc0789f4e7f48f2da09de4c76af343df8a6 100644
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -467,7 +467,7 @@ public class Sniffer extends Animal {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("snifferBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.popPush("snifferActivityUpdate");
         SnifferAi.updateActivity(this);
         profiler.pop();
diff --git a/net/minecraft/world/entity/monster/hoglin/Hoglin.java b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
index 288d03cd7fcf4d3e1536a0d692c31a0343063a8c..347fcd4bb181a7b8db763d52223555324083c048 100644
--- a/net/minecraft/world/entity/monster/hoglin/Hoglin.java
+++ b/net/minecraft/world/entity/monster/hoglin/Hoglin.java
@@ -136,7 +136,7 @@ public class Hoglin extends Animal implements Enemy, HoglinBase {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("hoglinBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         HoglinAi.updateActivity(this);
         if (this.isConverting()) {
diff --git a/net/minecraft/world/entity/monster/piglin/Piglin.java b/net/minecraft/world/entity/monster/piglin/Piglin.java
index e77384f3c7dd2cfe522ce2d91f71e4d6354c0a64..4044984936e64b486c0ddd2e3ff8928b31b90a12 100644
--- a/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -282,7 +282,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("piglinBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         PiglinAi.updateActivity(this);
         super.customServerAiStep(level);
diff --git a/net/minecraft/world/entity/monster/warden/Warden.java b/net/minecraft/world/entity/monster/warden/Warden.java
index eb7908467737e949ead0289bb43d40f008746fb8..c6cbe3c219ad333c8c6641645dae7b60ace654e6 100644
--- a/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/net/minecraft/world/entity/monster/warden/Warden.java
@@ -289,7 +289,7 @@ public class Warden extends Monster implements VibrationSystem {
     protected void customServerAiStep(final ServerLevel level) {
         ProfilerFiller profiler = Profiler.get();
         profiler.push("wardenBrain");
-        this.getBrain().tick(level, this);
+        if (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0) this.getBrain().tick(level, this);
         profiler.pop();
         super.customServerAiStep(level);
         if ((this.tickCount + this.getId()) % 120 == 0) {
diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index cba4af374ce9ae8f26770e399b8e6f97de69a287..7eb01cef4aaa5f3f72117f76a8dcc9f0977a5ebb 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -272,7 +272,7 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
         // Paper end - EAR 2
         ProfilerFiller profiler = Profiler.get();
         profiler.push("villagerBrain");
-        if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
+        if (!inactive && (this.activatedPriority <= 1 || this.tickCount % this.activatedPriority == 0)) this.getBrain().tick(level, this); // Paper - EAR 2
         profiler.pop();
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
