From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Frish2021 <1573880184@qq.com>
Date: Sun, 31 Aug 2025 13:13:19 +0800
Subject: [PATCH] Supported Scoreboard And Schedule Command.


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 3be546ea638dbde77196bbc2dc8a5b6a87e308c4..4f0d4ef09b73f378ac45609427a3e7d62d04a6c7 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -224,8 +224,8 @@ public class Commands {
         //RideCommand.register(this.dispatcher); // Folia - region threading - TODO later
         //RotateCommand.register(this.dispatcher); // Folia - region threading - TODO later
         SayCommand.register(this.dispatcher);
-        //ScheduleCommand.register(this.dispatcher); // Folia - region threading
-        //ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
+        if (dev.bacteriawa.mint.config.modules.experiment.ScheduleCommandConfig.enabled) ScheduleCommand.register(this.dispatcher); // Folia - region threading
+        if (dev.bacteriawa.mint.config.modules.experiment.ScoreboardCommandConfig.enabled) ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
         SeedCommand.register(this.dispatcher, selection != Commands.CommandSelection.INTEGRATED);
         VersionCommand.register(this.dispatcher, selection != Commands.CommandSelection.INTEGRATED);
         SetBlockCommand.register(this.dispatcher, context);
