From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Thu, 25 Dec 2025 15:13:08 +0800
Subject: [PATCH] Blank Configuration File


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index a1442b03a3506bf092ec76f32c0b033119a7d550..a0808c5af17b4bc8e6cbed4b7a1e577656e2e62e 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -201,7 +201,7 @@ public class Commands {
         //CloneCommands.register(this.dispatcher, context); // Folia - region threading - TODO
         DamageCommand.register(this.dispatcher, context);
         //DataCommands.register(this.dispatcher); // Folia - region threading - TODO
-        if (dev.bacteriawa.mint.config.modules.experiment.DatapackCommandConfig.enabled) DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
+        if (dev.bacteriawa.mint.config.modules.experiment.VanillaCommandConfig.datapack) DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
         //DebugCommand.register(this.dispatcher); // Folia - region threading - TODO
         DefaultGameModeCommands.register(this.dispatcher);
         //DialogCommand.register(this.dispatcher, context); // Folia - region threading - TODO
@@ -213,12 +213,12 @@ public class Commands {
         FillCommand.register(this.dispatcher, context);
         FillBiomeCommand.register(this.dispatcher, context);
         ForceLoadCommand.register(this.dispatcher);
-        if (dev.bacteriawa.mint.config.modules.experiment.DatapackCommandConfig.enabled) FunctionCommand.register(this.dispatcher); // Folia - region threading - TODO
+        if (dev.bacteriawa.mint.config.modules.experiment.VanillaCommandConfig.function) FunctionCommand.register(this.dispatcher); // Folia - region threading - TODO
         GameModeCommand.register(this.dispatcher);
         GameRuleCommand.register(this.dispatcher, context);
         GiveCommand.register(this.dispatcher, context);
         HelpCommand.register(this.dispatcher);
-        //ItemCommands.register(this.dispatcher, context); // Folia - region threading - TODO later
+        if (dev.bacteriawa.mint.config.modules.experiment.VanillaCommandConfig.item) ItemCommands.register(this.dispatcher, context); // Folia - region threading - TODO later
         KickCommand.register(this.dispatcher);
         KillCommand.register(this.dispatcher);
         ListPlayersCommand.register(this.dispatcher);
@@ -236,8 +236,8 @@ public class Commands {
         RideCommand.register(this.dispatcher);
         RotateCommand.register(this.dispatcher);
         SayCommand.register(this.dispatcher);
-        if (dev.bacteriawa.mint.config.modules.experiment.ScheduleCommandConfig.enabled) ScheduleCommand.register(this.dispatcher); // Folia - region threading
-        if (dev.bacteriawa.mint.config.modules.experiment.ScoreboardCommandConfig.enabled) ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
+        if (dev.bacteriawa.mint.config.modules.experiment.VanillaCommandConfig.schedule) ScheduleCommand.register(this.dispatcher); // Folia - region threading
+        if (dev.bacteriawa.mint.config.modules.experiment.VanillaCommandConfig.scoreboard) ScoreboardCommand.register(this.dispatcher, context); // Folia - region threading
         SeedCommand.register(this.dispatcher, selection != Commands.CommandSelection.INTEGRATED);
         VersionCommand.register(this.dispatcher, selection != Commands.CommandSelection.INTEGRATED);
         SetBlockCommand.register(this.dispatcher, context);
@@ -248,7 +248,7 @@ public class Commands {
         StopSoundCommand.register(this.dispatcher);
         StopwatchCommand.register(this.dispatcher);
         SummonCommand.register(this.dispatcher, context);
-        //TagCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        if(dev.bacteriawa.mint.config.modules.experiment.VanillaCommandConfig.tag) TagCommand.register(this.dispatcher); // Folia - region threading - TODO later
         //TeamCommand.register(this.dispatcher, context); // Folia - region threading - TODO later
         //TeamMsgCommand.register(this.dispatcher); // Folia - region threading - TODO later
         TeleportCommand.register(this.dispatcher);
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index df2823516d12f9528350f91e97a4c186f5ddd2d6..c71ae6e79483f4e384a74aa4c6cac745daa8d2f8 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -593,7 +593,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.paperConfigurations = services.paper().configurations(); // Paper - add paper configuration files
     }
 
-    protected abstract boolean initServer() throws IOException;
+    protected abstract boolean initServer() throws Exception;
 
     public ChunkLoadStatusView createChunkLoadStatusView(final int radius) {
         return new ChunkLoadStatusView() {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index d8faddc7392a405932eeb666639c75b8910dce6b..54c5bee70acd57b24cd09d57c74e427d39e6601a 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -165,7 +165,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
     }
 
     @Override
-    public boolean initServer() throws IOException {
+    public boolean initServer() throws Exception {
         int i = this.getProperties().managementServerPort;
         if (this.getProperties().managementServerEnabled) {
             String string = this.settings.getProperties().managementServerSecret;
@@ -261,6 +261,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
+        dev.bacteriawa.mint.core.MintBootstrap.setup();
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
