From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 24 Aug 2025 00:10:06 +0800
Subject: [PATCH] I18n support


diff --git a/net/minecraft/locale/Language.java b/net/minecraft/locale/Language.java
index 7b9e2a1a208b46a69c16e6afd8b502259893574f..adf3a2a362194ea1061ce0179bb68e16dafd11d0 100644
--- a/net/minecraft/locale/Language.java
+++ b/net/minecraft/locale/Language.java
@@ -65,7 +65,7 @@ public abstract class Language {
         };
     }
 
-    private static void parseTranslations(BiConsumer<String, String> output, String languagePath) {
+    public static void parseTranslations(BiConsumer<String, String> output, String languagePath) { // Mint - I18n support
         try (InputStream resourceAsStream = Language.class.getResourceAsStream(languagePath)) {
             loadFromJson(resourceAsStream, output);
         } catch (JsonParseException | IOException var7) {
diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index a6d11446fe5791254047b88076371f801b9d4435..50fd46133b429837008ac1929e5a530f20265dd6 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -18,6 +18,7 @@ import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
 
 import dev.bacteriawa.mint.config.MintConfig;
+import dev.bacteriawa.mint.utils.ServerI18nUtil;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.OptionSpec;
@@ -144,6 +145,8 @@ public class Main {
                 return;
             }
 
+            ServerI18nUtil.init(); // Mint - I18n support
+
             // Spigot start
             boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
             if (eulaAgreed) {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 8da291c8171ee453cba13fbe8fde50fbc4c4b40d..e07998cf02c3ee5099cdfebde0cd6134c3441c76 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -27,6 +27,7 @@ import dev.bacteriawa.mint.config.MintConfig;
 import dev.bacteriawa.mint.functions.GlobalServerMemoryBar;
 import dev.bacteriawa.mint.functions.GlobalServerRegionBar;
 import dev.bacteriawa.mint.functions.GlobalServerTpsBar;
+import dev.bacteriawa.mint.utils.ServerI18nUtil;
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
 import net.minecraft.SharedConstants;
@@ -181,6 +182,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
         MintConfig.setup(); // Mint
+        ServerI18nUtil.preInit(); // Mint - I18n support
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
