From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 4 Feb 2026 21:54:45 +0800
Subject: [PATCH] HMBIRD scheduler


diff --git a/io/papermc/paper/threadedregions/TickRegionScheduler.java b/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 439c5a222fcf51c9dc5bfef9cdb23f419fffc28c..3b0273e3f8511b53fe37e6e36f957ee22d5edb85 100644
--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -2,6 +2,7 @@ package io.papermc.paper.threadedregions;
 
 import ca.spottedleaf.concurrentutil.numa.OSNuma;
 import ca.spottedleaf.concurrentutil.scheduler.EDFSchedulerThreadPool;
+import dev.bacteriawa.mint.scheduler.HMBIRDSchedulerThreadPool;
 import ca.spottedleaf.concurrentutil.scheduler.SchedulableTick;
 import ca.spottedleaf.concurrentutil.scheduler.Scheduler;
 import ca.spottedleaf.concurrentutil.scheduler.StealingScheduledThreadPool;
@@ -18,6 +19,7 @@ import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.ChunkPos;
 import org.slf4j.Logger;
+import dev.bacteriawa.mint.config.modules.scheduler.HmbirdSchedulerConfig;
 import java.lang.management.ManagementFactory;
 import java.lang.management.ThreadMXBean;
 import java.util.concurrent.ThreadFactory;
@@ -53,6 +55,7 @@ public final class TickRegionScheduler {
 
     public static enum SchedulerType {
         EDF,
+        HMBIRD,
         WORK_STEALING;
     }
 
@@ -77,6 +80,13 @@ public final class TickRegionScheduler {
                 this.scheduler = new EDFSchedulerThreadPool(initialThreads, threadFactory);
                 break;
             }
+            case HMBIRD: {
+                final HMBIRDSchedulerThreadPool hmbird = new HMBIRDSchedulerThreadPool(initialThreads, threadFactory);
+                hmbird.setIntermediateTimeSliceNs(TimeUnit.MILLISECONDS.toNanos(HmbirdSchedulerConfig.intermediateTimeSliceMs));
+                hmbird.setQueueRejectThresholds(HmbirdSchedulerConfig.rejectGlobalQueueSize, HmbirdSchedulerConfig.recoverGlobalQueueSize);
+                this.scheduler = hmbird;
+                break;
+            }
             case WORK_STEALING: {
                 this.scheduler = new StealingScheduledThreadPool(
                         threadFactory, MoonriseConstants.NUMA_ENABLE ? OSNuma.getNativeInstance() : OSNuma.NoOp.INSTANCE
@@ -94,6 +104,9 @@ public final class TickRegionScheduler {
         if (this.scheduler instanceof EDFSchedulerThreadPool edfSchedulerThreadPool) {
             edfSchedulerThreadPool.start();
         }
+        if (this.scheduler instanceof HMBIRDSchedulerThreadPool hmbirdSchedulerThreadPool) {
+            hmbirdSchedulerThreadPool.start();
+        }
     }
 
     public void setThreads(final int threads) {
