From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Frish2021 <1573880184@qq.com>
Date: Sun, 16 Nov 2025 13:56:55 +0800
Subject: [PATCH] Decoupled Configuration System.


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 51b000de7b4cef098cfbcfdb4273318d9dfb378c..58b9f5606a0ef8bf3194dfaa2b34e94dcc6e1f4b 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -16,9 +16,7 @@ import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
 
-import dev.bacteriawa.mint.utils.ServerI18nUtil;
 import joptsimple.OptionSet;
-import dev.bacteriawa.mint.config.MintConfig;
 import net.minecraft.CrashReport;
 import net.minecraft.SharedConstants;
 import net.minecraft.SuppressForbidden;
@@ -107,8 +105,6 @@ public class Main {
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
 
-//            MintConfig.loadConfig(); // Mint
-            MintConfig.loadAllConfigs(); // Mint - new config system.
             io.papermc.paper.plugin.PluginInitializerManager.load(optionSet); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
@@ -141,8 +137,6 @@ public class Main {
                 return;
             }
 
-            ServerI18nUtil.init(); // Mint - I18n support
-
             // Spigot start
             boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
             if (eulaAgreed) {
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index af713e4f17355f4c6976ec2897686bdf035b0ff0..295e16f64cce561a6cca0478b5d8014af1027d2e 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -504,7 +504,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         dataStorage.computeIfAbsent(ServerScoreboard.TYPE);
     }
 
-    protected abstract boolean initServer() throws IOException;
+    protected abstract boolean initServer() throws Exception;
 
     protected void loadLevel(String levelId) { // CraftBukkit
         if (!JvmProfiler.INSTANCE.isRunning()) {
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index b1ba88212535907d09ac6664f90ee49b30743f9d..a60d57d9dea68f8aeb5fd6d385b0b8b764f52ac5 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -17,9 +17,6 @@ import java.util.Locale;
 import java.util.Optional;
 import javax.annotation.Nullable;
 
-import dev.bacteriawa.mint.exception.MintRuntimeException;
-import dev.bacteriawa.mint.utils.ServerI18nUtil;
-import dev.bacteriawa.mint.config.MintConfig;
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
 import net.minecraft.SharedConstants;
@@ -100,7 +97,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
     }
 
     @Override
-    public boolean initServer() throws IOException {
+    public boolean initServer() throws Exception {
         Thread thread = new Thread("Server console handler") {
             @Override
             public void run() {
@@ -172,13 +169,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
-//        MintConfig.setup(); // Mint
-        try {
-            MintConfig.setupAllConfigs(); // Mint - new config system.
-        } catch (InvocationTargetException | IllegalAccessException e) {
-            throw new MintRuntimeException(e);
-        }
-        ServerI18nUtil.preInit(); // Mint - I18n support
+        dev.bacteriawa.mint.core.MintBootstrap.setup();
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
