From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Tue, 11 Feb 2025 09:16:49 +0800
Subject: [PATCH] KioCG Chunk API


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 9f8414f854646ad7429fe806e35a13e32a07ddaa..9a0995bf82830b45b2d72b2a1cff2bfdd4f7a638 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -388,4 +388,6 @@ public interface Chunk extends PersistentDataHolder {
          */
         UNLOADED;
     }
+
+    long getChunkHotAvg(); // KioCG
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 672d78811b27d046bea2fb60a5cd8da387d37b92..3ae8cad7c1ab0ef3a5c228492ea9d4a2f149b943 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3932,4 +3932,6 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @ApiStatus.Experimental
     PlayerGameConnection getConnection();
+
+    long getNearbyChunkHot(); // KioCG
 }
