From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 25 May 2025 01:22:39 +0800
Subject: [PATCH] Remove stream for EntityDamageEvent


diff --git a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
index 6d5ae34b39d6001effc2faadd6c30e3a8ab109fa..485eae9a9e7fa166a76d2328cec2cc2dc1e15810 100644
--- a/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
+++ b/src/main/java/org/bukkit/event/entity/EntityDamageEvent.java
@@ -57,9 +57,13 @@ public class EntityDamageEvent extends EntityEvent implements Cancellable {
         super(damagee);
         Preconditions.checkArgument(modifiers.containsKey(DamageModifier.BASE), "BASE DamageModifier missing");
         Preconditions.checkArgument(!modifiers.containsKey(null), "Cannot have null DamageModifier");
-        Preconditions.checkArgument(modifiers.values().stream().allMatch(Objects::nonNull), "Cannot have null modifier values");
+        for (Double value : modifiers.values()) {
+            Preconditions.checkArgument(value != null, "Cannot have null modifier values");
+        }
         Preconditions.checkArgument(modifiers.keySet().equals(modifierFunctions.keySet()), "Must have a modifier function for each DamageModifier");
-        Preconditions.checkArgument(modifierFunctions.values().stream().allMatch(Objects::nonNull), "Cannot have null modifier function");
+        for (Function<? super Double, Double> function : modifierFunctions.values()) {
+            Preconditions.checkArgument(function != null, "Cannot have null modifier function");
+        }
         this.originals = new EnumMap<>(modifiers);
         this.cause = cause;
         this.modifiers = modifiers;
