From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Tue, 8 Jul 2025 23:01:59 +0800
Subject: [PATCH] Refactoring FoliaScheduler


diff --git a/src/main/java/io/papermc/paper/threadedregions/scheduler/ScheduledTask.java b/src/main/java/io/papermc/paper/threadedregions/scheduler/ScheduledTask.java
index a6b50c9d8af589cc4747e14d343d2045216c249c..3eca64c7e16e7716c10da40297e7510e25ce9d1a 100644
--- a/src/main/java/io/papermc/paper/threadedregions/scheduler/ScheduledTask.java
+++ b/src/main/java/io/papermc/paper/threadedregions/scheduler/ScheduledTask.java
@@ -76,7 +76,7 @@ public interface ScheduledTask {
          * The repeating task's future executions are cancelled already, but the task is currently
          * being executed.
          */
-        NEXT_RUNS_CANCELLED_ALREADY,
+        NEXT_RUNS_CANCELLED_ALREADY, UNCERTAIN,
     }
 
     /**
@@ -107,6 +107,6 @@ public interface ScheduledTask {
         /**
          * The task is repeating and currently executing, but future executions are cancelled and will not occur.
          */
-        CANCELLED_RUNNING;
+        CANCELLED_RUNNING, UNKNOWN;
     }
 }
