From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Thu, 29 May 2025 01:05:10 +0800
Subject: [PATCH] Add tpsbar with chunkhot membar and regionbar


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index 400e632208d133a3f49fc7f14bceb48a1026769b..631bec0adee5b01bfb931c25195b949eaf2efd27 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -422,4 +422,12 @@ public class CraftChunk implements Chunk {
     static {
         Arrays.fill(FULL_LIGHT, (byte) 0xFF);
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getChunkHotAvg() {
+        final net.minecraft.world.level.chunk.LevelChunk target = this.level.getChunkIfLoaded(this.x,this.z);
+        return target == null ? -1 : target.getChunkHot().getAverage();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0e32a9ad852e521ce9618acc1642d3748995b4bf..6e442a8ecd5039884d9158a491adf44e9e5f3e1f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2368,6 +2368,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             handle.newLevel = data.getIntOr("newLevel", 0);
             handle.expToDrop = data.getIntOr("expToDrop", 0);
             handle.keepLevel = data.getBooleanOr("keepLevel", false);
+            //Luminol start - Tpsbar
+            getHandle().isTpsBarVisible = data.getBooleanOr("tpsbarVisible", false);
+            //Luminol end
+            //Luminol start - Membar
+            getHandle().isMemBarVisible = data.getBooleanOr("membarVisible", false);
+            //Luminol end
+            //Luminol start - Regionbar
+            getHandle().isRegionBarVisible = data.getBooleanOr("regionbarVisible", false);
+            //Luminol end
         });
     }
 
@@ -2389,6 +2398,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         data.putLong("lastPlayed", System.currentTimeMillis());
         data.putString("lastKnownName", handle.getScoreboardName());
 
+        //Luminol start - Tpsbar
+        data.putBoolean("tpsbarVisible",handle.isTpsBarVisible);
+        //Luminol end
+        //Luminol start - Membar
+        data.putBoolean("membarVisible", handle.isMemBarVisible);
+        //Luminol end
+        //Luminol start - Regionbar
+        data.putBoolean("regionbarVisible", handle.isRegionBarVisible);
+        //Luminol end
         // Paper start - persist for use in offline save data
         if (!tag.contains("Paper")) {
             tag.put("Paper", new CompoundTag());
@@ -3581,4 +3599,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
     public void setDeathScreenScore(final int score) {
         getHandle().setScore(score);
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getNearbyChunkHot() {
+        return this.getHandle().getNearbyChunkHot();
+    }
+    // KioCG end
 }
