From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 20 Dec 2025 02:57:12 +0800
Subject: [PATCH] Add tpsbar with chunkhot membar and regionbar


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index a9712c50cb8002d5b63191701f6b1832369857e4..edbfd718242d4e8ba684e0231959f31cf007d75e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -413,4 +413,12 @@ public class CraftChunk implements Chunk {
     static {
         Arrays.fill(FULL_LIGHT, (byte) 0xFF);
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getChunkHotAvg() {
+        final net.minecraft.world.level.chunk.LevelChunk target = this.level.getChunkIfLoaded(this.x,this.z);
+        return target == null ? -1 : target.getChunkHot().getAverage();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 0ab8a2d8c41b22df271e342ba341d5d1cd7cef7f..4fb313607aaeb6fa1cfae0fdf22ff080eeeb9273 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2204,6 +2204,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             handle.newLevel = data.getIntOr("newLevel", 0);
             handle.expToDrop = data.getIntOr("expToDrop", 0);
             handle.keepLevel = data.getBooleanOr("keepLevel", false);
+            //Luminol start - Tpsbar
+            getHandle().isTpsBarVisible = data.getBooleanOr("tpsbarVisible", false);
+            //Luminol end
+            //Luminol start - Membar
+            getHandle().isMemBarVisible = data.getBooleanOr("membarVisible", false);
+            //Luminol end
+            //Luminol start - Regionbar
+            getHandle().isRegionBarVisible = data.getBooleanOr("regionbarVisible", false);
+            //Luminol end
         });
     }
 
@@ -2221,6 +2230,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         data.putLong("lastPlayed", System.currentTimeMillis());
         data.putString("lastKnownName", handle.getGameProfile().name());
 
+        //Luminol start - Tpsbar
+        data.putBoolean("tpsbarVisible",handle.isTpsBarVisible);
+        //Luminol end
+        //Luminol start - Membar
+        data.putBoolean("membarVisible", handle.isMemBarVisible);
+        //Luminol end
+        //Luminol start - Regionbar
+        data.putBoolean("regionbarVisible", handle.isRegionBarVisible);
+        //Luminol end
         // Paper start - persist for use in offline save data
         ValueOutput paper = output.child("Paper");
         paper.putLong("LastLogin", handle.loginTime);
@@ -3405,4 +3423,10 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             entity.remove();
         }
     }
+    // KioCG start - ChunkHot
+    @Override
+    public long getNearbyChunkHot() {
+        return this.getHandle().getNearbyChunkHot();
+    }
+    // KioCG end
 }
