From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 14 Jul 2025 16:58:35 +0800
Subject: [PATCH] Add tpsbar with chunkhot membar and regionbar


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index c65b26d22fb72300e487ac6d1a6d7ed781a13d99..90718b8958399d570343c3a21da5966aed765f12 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -427,4 +427,12 @@ public class CraftChunk implements Chunk {
     static {
         Arrays.fill(FULL_LIGHT, (byte) 0xFF);
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getChunkHotAvg() {
+        final net.minecraft.world.level.chunk.LevelChunk target = this.level.getChunkIfLoaded(this.x,this.z);
+        return target == null ? -1 : target.getChunkHot().getAverage();
+    }
+    // KioCG end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 4cdbd0965f42f30979a24444625532a29b43f9c3..c1052533b0debf931d5857ba8322255b3543089d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2442,6 +2442,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             handle.newLevel = data.getIntOr("newLevel", 0);
             handle.expToDrop = data.getIntOr("expToDrop", 0);
             handle.keepLevel = data.getBooleanOr("keepLevel", false);
+            //Luminol start - Tpsbar
+            getHandle().isTpsBarVisible = data.getBooleanOr("tpsbarVisible", false);
+            //Luminol end
+            //Luminol start - Membar
+            getHandle().isMemBarVisible = data.getBooleanOr("membarVisible", false);
+            //Luminol end
+            //Luminol start - Regionbar
+            getHandle().isRegionBarVisible = data.getBooleanOr("regionbarVisible", false);
+            //Luminol end
         });
     }
 
@@ -2459,6 +2468,15 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         data.putLong("lastPlayed", System.currentTimeMillis());
         data.putString("lastKnownName", handle.getScoreboardName());
 
+        //Luminol start - Tpsbar
+        data.putBoolean("tpsbarVisible",handle.isTpsBarVisible);
+        //Luminol end
+        //Luminol start - Membar
+        data.putBoolean("membarVisible", handle.isMemBarVisible);
+        //Luminol end
+        //Luminol start - Regionbar
+        data.putBoolean("regionbarVisible", handle.isRegionBarVisible);
+        //Luminol end
         // Paper start - persist for use in offline save data
         ValueOutput paper = output.child("Paper");
         paper.putLong("LastLogin", handle.loginTime);
@@ -3606,4 +3624,11 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             this(new ShortArraySet(), new ArrayList<>());
         }
     }
+
+    // KioCG start - ChunkHot
+    @Override
+    public long getNearbyChunkHot() {
+        return this.getHandle().getNearbyChunkHot();
+    }
+    // KioCG end
 }
