From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sat, 24 May 2025 02:26:33 +0800
Subject: [PATCH] Leaves: Fix SculkCatalyst exp skip


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 00416e0fe016b1e6502a0b57ffc623c68e8f2893..e040f9a7c9451fbe547dda56952410047a99ddc0 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -862,7 +862,7 @@ public class CraftEventFactory {
         CraftLivingEntity entity = (CraftLivingEntity) victim.getBukkitEntity();
         CraftDamageSource bukkitDamageSource = new CraftDamageSource(damageSource);
         CraftWorld world = (CraftWorld) entity.getWorld();
-        EntityDeathEvent event = new EntityDeathEvent(entity, bukkitDamageSource, new io.papermc.paper.util.TransformingRandomAccessList<>(drops, Entity.DefaultDrop::stack, FROM_FUNCTION), victim.getExpReward(world.getHandle(), damageSource.getEntity())); // Paper - Restore vanilla drops behavior
+        EntityDeathEvent event = new EntityDeathEvent(entity, bukkitDamageSource, new io.papermc.paper.util.TransformingRandomAccessList<>(drops, Entity.DefaultDrop::stack, FROM_FUNCTION), victim.getExpReward(world.getHandle(), damageSource.getEntity()), victim.getExperienceReward(world.getHandle(), damageSource.getEntity())); // Paper - Restore vanilla drops behavior // Leaves - exp fix
         populateFields(victim, event); // Paper - make cancellable
         Bukkit.getServer().getPluginManager().callEvent(event);
 
@@ -873,6 +873,7 @@ public class CraftEventFactory {
         playDeathSound(victim, event, damageSource);
         // Paper end
         victim.expToDrop = event.getDroppedExp();
+        victim.expToReward = event.getRewardExp(); // Leaves - exp fix
         lootCheck.run(); // Paper - advancement triggers before destroying items
 
         // Paper start - Restore vanilla drops behavior
@@ -908,6 +909,7 @@ public class CraftEventFactory {
         victim.newLevel = event.getNewLevel();
         victim.newTotalExp = event.getNewTotalExp();
         victim.expToDrop = event.getDroppedExp();
+        victim.expToReward = event.getRewardExp(); // Leaves - exp fix
         victim.newExp = event.getNewExp();
 
         // Paper start - Restore vanilla drops behavior
