From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 6 Aug 2025 00:08:53 +0800
Subject: [PATCH] Add Networkbar


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 889af1ab7c378cf4fe97d8af2cf5e8ee09ca9c20..6b667a0a04dd5454abb0ef93ff6bddfded3a13a4 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2213,6 +2213,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             //Luminol start - Regionbar
             getHandle().isRegionBarVisible = data.getBooleanOr("regionbarVisible", false);
             //Luminol end
+            // Mint start - Networkbar
+            getHandle().isNetworkBarVisible = data.getBooleanOr("networkBarVisible", false);
+            // Mint end - Networkbar
         });
     }
 
@@ -2239,6 +2242,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         //Luminol start - Regionbar
         data.putBoolean("regionbarVisible", handle.isRegionBarVisible);
         //Luminol end
+        // Mint start - Networkbar
+        data.putBoolean("networkbarVisible", handle.isNetworkBarVisible);
+        // Mint end - Networkbar
         // Paper start - persist for use in offline save data
         ValueOutput paper = output.child("Paper");
         paper.putLong("LastLogin", handle.loginTime);
