From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 6 Aug 2025 00:08:53 +0800
Subject: [PATCH] Add Networkbar


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 75a591e4357fb51d1e28b1811972061203a613fe..81fd7f66e5b2c3e8e6b748a52e75d4279d1bd189 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2207,6 +2207,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
             //Luminol start - Regionbar
             getHandle().isRegionBarVisible = data.getBooleanOr("regionbarVisible", false);
             //Luminol end
+            // Mint start - Networkbar
+            getHandle().isNetworkBarVisible = data.getBooleanOr("networkBarVisible", false);
+            // Mint end - Networkbar
         });
     }
 
@@ -2233,6 +2236,9 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         //Luminol start - Regionbar
         data.putBoolean("regionbarVisible", handle.isRegionBarVisible);
         //Luminol end
+        // Mint start - Networkbar
+        data.putBoolean("networkbarVisible", handle.isNetworkBarVisible);
+        // Mint end - Networkbar
         // Paper start - persist for use in offline save data
         ValueOutput paper = output.child("Paper");
         paper.putLong("LastLogin", handle.loginTime);
