From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 23 Feb 2025 17:20:51 +0800
Subject: [PATCH] KioCG Chunk API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fac4aec289e07231d80a9890653432f688355afa..40d6bc7c4f39381de97d97bcc8c863df080236da 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3603,6 +3603,8 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     @Deprecated(since = "1.20.1")
     void removeAdditionalChatCompletions(@NotNull java.util.Collection<String> completions);
+
+    long getNearbyChunkHot(); // KioCG
     // Paper end - custom chat completions API
 
     // Spigot start
