From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xuyin <1573880184@qq.com>
Date: Thu, 16 Jan 2025 18:12:15 +0800
Subject: [PATCH] Removed some timing.


diff --git a/src/main/java/org/bukkit/plugin/PluginManager.java b/src/main/java/org/bukkit/plugin/PluginManager.java
index 47153dee66782a00b980ecf15e8774ab6f3d887d..e60617ee040198b16338d433efcfa1a42630206e 100644
--- a/src/main/java/org/bukkit/plugin/PluginManager.java
+++ b/src/main/java/org/bukkit/plugin/PluginManager.java
@@ -310,9 +310,13 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
     /**
      * Returns whether or not timing code should be used for event calls
      *
+     * @deprecated This method will be removed in the future
      * @return True if event timings are to be used
      */
-    public boolean useTimings();
+    @Deprecated(forRemoval = true)
+    default boolean useTimings() {
+        throw new UnsupportedOperationException();
+    };
 
     // Paper start
     @org.jetbrains.annotations.ApiStatus.Internal
diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index 065352b383d6398382b4f14641859dbfe2e6a355..df8a97b96931ae641fcf53b5b008793887c627c2 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -955,8 +955,7 @@ public final class SimplePluginManager implements PluginManager {
 
     @Override
     public boolean useTimings() {
-        if (true) {return this.paperPluginManager.useTimings();} // Paper
-        return co.aikar.timings.Timings.isTimingsEnabled(); // Spigot
+        return false;
     }
 
     /**
diff --git a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
index def75eab04c412db9b7c4b5227ee547696520846..3235657f0b2d78dda940bba5e9461ea4ff4334e2 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPluginLoader.java
@@ -231,7 +231,6 @@ public final class JavaPluginLoader implements PluginLoader {
         Preconditions.checkArgument(plugin != null, "Plugin can not be null");
         Preconditions.checkArgument(listener != null, "Listener can not be null");
 
-        boolean useTimings = server.getPluginManager().useTimings();
         Map<Class<? extends Event>, Set<RegisteredListener>> ret = new HashMap<Class<? extends Event>, Set<RegisteredListener>>();
         Set<Method> methods;
         try {
