From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: xuyin <1573880184@qq.com>
Date: Sat, 11 Jan 2025 11:45:22 +0800
Subject: [PATCH] Fixed reload command.


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 5df19bd701c67506689fc7f49d91f99ebfbc83f0..b88fa0f8e42134f8c0d00f45757f12613c1faf04 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -179,7 +179,12 @@ public class SimpleCommandMap implements CommandMap {
         return true;
     }
 
+    /**
+     * @deprecated I don`t know why this failed to reload commands.
+     */
+    // Mint start - Fixed reload command.
     @Override
+    @Deprecated
     public synchronized void clearCommands() {
         for (Map.Entry<String, Command> entry : knownCommands.entrySet()) {
             entry.getValue().unregister(this);
@@ -187,6 +192,7 @@ public class SimpleCommandMap implements CommandMap {
         knownCommands.clear();
         setDefaultCommands();
     }
+    // Mint end - Fixed reload command.
 
     @Override
     @Nullable
