From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 22 Mar 2025 02:33:27 +0800
Subject: [PATCH] Add config for unsafe teleportation


diff --git a/src/main/java/dev/bacteriawa/mint/config/modules/fixes/UnsafeTeleportationConfig.java b/src/main/java/dev/bacteriawa/mint/config/modules/fixes/UnsafeTeleportationConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..df0ad8402750419eaa58307f00a8b5871eeda2b1
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/config/modules/fixes/UnsafeTeleportationConfig.java
@@ -0,0 +1,14 @@
+package dev.bacteriawa.mint.config.modules.fixes;
+
+import dev.bacteriawa.mint.config.ConfigCategory;
+import dev.bacteriawa.mint.config.ConfigField;
+import dev.bacteriawa.mint.config.Configuration;
+
+@Configuration(name = "unsafe-teleportation", type = ConfigCategory.fixes)
+public class UnsafeTeleportationConfig {
+    @ConfigField(comment = {
+            "If you want to use sand duping,please turn on this.",
+            "Warning: This would cause some unsafe issues, you could learn more on : https://github.com/PaperMC/Folia/issues/297"
+    })
+    public static boolean enabled = false;
+}
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index a498cdbe42df3161c35b2a1652dfa63693366bd6..895654e706ba70822afb92fe4b033901b074f4f0 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity.item;
 
+import dev.bacteriawa.mint.config.modules.fixes.UnsafeTeleportationConfig;
 import com.mojang.logging.LogUtils;
 import java.util.Iterator;
 import java.util.function.Predicate;
@@ -69,7 +70,7 @@ public class FallingBlockEntity extends Entity {
     public float fallDamagePerDistance;
     @Nullable
     public CompoundTag blockData;
-    public boolean forceTickAfterTeleportToDuplicate;
+    public boolean forceTickAfterTeleportToDuplicate = UnsafeTeleportationConfig.enabled;
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
 
@@ -430,7 +431,7 @@ public class FallingBlockEntity extends Entity {
         boolean flag = (resourcekey1 == Level.END || resourcekey == Level.END) && resourcekey1 != resourcekey;
         Entity entity = super.teleport(teleportTarget);
 
-        this.forceTickAfterTeleportToDuplicate = entity != null && flag && io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowUnsafeEndPortalTeleportation; // Paper
+        this.forceTickAfterTeleportToDuplicate = entity != null && flag && (io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowUnsafeEndPortalTeleportation || UnsafeTeleportationConfig.enabled); // Paper
         return entity;
     }
 }
