From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Wed, 22 Jan 2025 18:31:34 +0800
Subject: [PATCH] Add config for server mod name


diff --git a/src/main/java/dev/bacteriawa/mint/config/modules/misc/ServerBrandNameConfig.java b/src/main/java/dev/bacteriawa/mint/config/modules/misc/ServerBrandNameConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..fdb59e29b9da6bdf3bd363765bcb58543e34dc13
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/config/modules/misc/ServerBrandNameConfig.java
@@ -0,0 +1,11 @@
+package dev.bacteriawa.mint.config.modules.misc;
+
+import dev.bacteriawa.mint.config.ConfigCategory;
+import dev.bacteriawa.mint.config.ConfigField;
+import dev.bacteriawa.mint.config.Configuration;
+
+@Configuration(name = "server-brand-name", type = ConfigCategory.misc)
+public class ServerBrandNameConfig {
+    @ConfigField(comment = "This can change the server name.")
+    public static String serverModName= "Mint";
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1676c4abb3f6f71bb7b25351aa58b4e127786fca..d84f644bf308b0087f370306e8ba76f0f560ffb6 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2121,7 +2121,10 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        if(false)
+            return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+
+        return dev.bacteriawa.mint.config.modules.misc.ServerBrandNameConfig.serverModName; // Mint
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
