From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kercute <A3167717663@hotmail.com>
Date: Sat, 18 Jan 2025 13:59:38 +0800
Subject: [PATCH] Add config for unsafe teleportation


diff --git a/src/main/java/com/menthamc/mint/config/modules/fix/UnsafeTeleportationConfig.java b/src/main/java/com/menthamc/mint/config/modules/fix/UnsafeTeleportationConfig.java
index 6fee15142ee604a094b71a743fe004063536e499..9a1e4c5b27cbbce4cc495ad60a79f73477427b12 100644
--- a/src/main/java/com/menthamc/mint/config/modules/fix/UnsafeTeleportationConfig.java
+++ b/src/main/java/com/menthamc/mint/config/modules/fix/UnsafeTeleportationConfig.java
@@ -1,15 +1,13 @@
 package com.menthamc.mint.config.modules.fix;
 
 import com.menthamc.mint.config.ConfigCategory;
-import com.menthamc.mint.config.ConfigField;
+import com.menthamc.mint.config.ConfigInfo;
 import com.menthamc.mint.config.Configuration;
 import com.menthamc.mint.config.IConfigModule;
 
-@Configuration(name = "unsafe-teleportation", type = ConfigCategory.fix)
+@Configuration(name = "unsafe-teleportation", type = ConfigCategory.FIX)
 public class UnsafeTeleportationConfig implements IConfigModule {
-    @ConfigField(comment = {
-            "If you want to use sand duping,please turn on this.",
-            "Warning: This would cause some unsafe issues, you could learn more on : https://github.com/PaperMC/Folia/issues/297"
-    })
-    public static boolean enabled = false;
+    @ConfigInfo(commit = "If you want to use sand duping,please turn on this.\n" +
+            "Warning: This would cause some unsafe issues, you could learn more on : https://github.com/PaperMC/Folia/issues/297")
+    public boolean enabled = false;
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index c3940456f1e0f7914bbd6ed01d8e5a1edf62e916..6969b31173cd8499c7f203eda2616550e5539407 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity.item;
 
+import com.menthamc.mint.config.IConfigModule;
 import com.menthamc.mint.config.modules.fix.UnsafeTeleportationConfig;
 import com.mojang.logging.LogUtils;
 import java.util.Iterator;
@@ -70,7 +71,7 @@ public class FallingBlockEntity extends Entity {
     public float fallDamagePerDistance;
     @Nullable
     public CompoundTag blockData;
-    public static boolean forceTickAfterTeleportToDuplicate =  UnsafeTeleportationConfig.enabled; // Mint
+    public static boolean forceTickAfterTeleportToDuplicate =  IConfigModule.get(UnsafeTeleportationConfig.class).enabled; // Mint
     protected static final EntityDataAccessor<BlockPos> DATA_START_POS = SynchedEntityData.defineId(FallingBlockEntity.class, EntityDataSerializers.BLOCK_POS);
     public boolean autoExpire = true; // Paper - Expand FallingBlock API
 
diff --git a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
index e1c071e0b25c4678bf08313b6c60cf895a54aa02..1f23a09a16a33e3a5a026abd64eaa5da0b954b5e 100644
--- a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.level.block;
 
+import static net.minecraft.world.entity.item.FallingBlockEntity.forceTickAfterTeleportToDuplicate;
 import com.mojang.serialization.MapCodec;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -72,6 +73,11 @@ public class EndPortalBlock extends BaseEntityBlock implements Portal {
             // CraftBukkit end
             // Folia - region threading - do not show credits
 
+            // Mint Start
+            if (forceTickAfterTeleportToDuplicate && !(entity instanceof net.minecraft.world.entity.player.Player)) {
+                entity.endPortalLogicAsync(pos);
+            }
+            // Mint End
             entity.setAsInsidePortal(this, pos);
         }
 
