From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kercute <A3167717663@hotmail.com>
Date: Sat, 18 Jan 2025 18:13:22 +0800
Subject: [PATCH] Add config for server mod name


diff --git a/src/main/java/com/menthamc/mint/config/ConfigCategory.java b/src/main/java/com/menthamc/mint/config/ConfigCategory.java
index 2e91061279e8efb96ef458d1213ff8b054001592..b77447ff13e627a5537003ed647c74cf33c31dd1 100644
--- a/src/main/java/com/menthamc/mint/config/ConfigCategory.java
+++ b/src/main/java/com/menthamc/mint/config/ConfigCategory.java
@@ -1,6 +1,12 @@
 package com.menthamc.mint.config;
 
 public enum ConfigCategory {
-    fix,
-    misc;
+    FIX("fix"),
+    MISC("misc");
+
+    final String name;
+
+    ConfigCategory(String name) {
+        this.name = name;
+    }
 }
diff --git a/src/main/java/com/menthamc/mint/config/modules/misc/ServerModNameConfig.java b/src/main/java/com/menthamc/mint/config/modules/misc/ServerModNameConfig.java
index 40d274ced66349840a1cb391a5d2a8f5f49f7b17..ae7cf97d0932e93cf814647f4847bbc8eaf3a70e 100644
--- a/src/main/java/com/menthamc/mint/config/modules/misc/ServerModNameConfig.java
+++ b/src/main/java/com/menthamc/mint/config/modules/misc/ServerModNameConfig.java
@@ -1,12 +1,12 @@
 package com.menthamc.mint.config.modules.misc;
 
 import com.menthamc.mint.config.ConfigCategory;
-import com.menthamc.mint.config.ConfigField;
+import com.menthamc.mint.config.ConfigInfo;
 import com.menthamc.mint.config.Configuration;
 import com.menthamc.mint.config.IConfigModule;
 
-@Configuration(name = "server-name", type = ConfigCategory.misc)
+@Configuration(name = "server-name", type = ConfigCategory.MISC)
 public class ServerModNameConfig implements IConfigModule {
-    @ConfigField(comment = "This can change the server name.")
+    @ConfigInfo(commit = "This can change the server name.")
     public static String serverModName= "Mint";
 }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index e067b2950baef8346c9c4bf85b0eb073af8b1bee..19fea0491864e60f3ebbeed3318e638bbddc5060 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2028,7 +2028,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return com.menthamc.mint.config.modules.misc.ServerModNameConfig.serverModName; // Mint
     }
 
     public SystemReport fillSystemReport(SystemReport details) {
