From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 22 Mar 2025 02:38:15 +0800
Subject: [PATCH] SIMD utilities


diff --git a/build.gradle.kts b/build.gradle.kts
index 0937936d5832b89e221c9d3ea054a3fc30545a97..fbf81d9ebd9c05f31329d2569c58bdfb082dded8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -89,6 +89,13 @@ paperweight {
     craftBukkitPackageVersion.set("v1_21_R3") // also needs to be updated in MappingEnvironment
 }
 
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
+
 tasks.jar {
     archiveClassifier.set("dev")
 
diff --git a/src/main/java/dev/bacteriawa/mint/config/modules/optimizations/SIMDConfig.java b/src/main/java/dev/bacteriawa/mint/config/modules/optimizations/SIMDConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..819326b86cc0a8b39d79673545d34d2c91bc1d25
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/config/modules/optimizations/SIMDConfig.java
@@ -0,0 +1,35 @@
+package dev.bacteriawa.mint.config.modules.optimizations;
+
+import com.electronwill.nightconfig.core.file.CommentedFileConfig;
+import dev.bacteriawa.mint.config.ConfigCategory;
+import dev.bacteriawa.mint.config.Configuration;
+import com.mojang.logging.LogUtils;
+import gg.pufferfish.pufferfish.simd.SIMDDetection;
+import org.slf4j.Logger;
+
+@Configuration(name = "simd", type = ConfigCategory.optimisations)
+public class SIMDConfig {
+    private static final Logger LOGGER = LogUtils.getLogger();
+    public static boolean enabled = true;
+
+    public void loaded(CommentedFileConfig Config) {
+        try {
+            SIMDDetection.isEnabled = SIMDDetection.canEnable(LOGGER);
+            SIMDDetection.versionLimited = SIMDDetection.getJavaVersion() < 17;
+        } catch (NoClassDefFoundError | Exception ignored) {
+            ignored.printStackTrace();
+        }
+
+        if (SIMDDetection.isEnabled) {
+            LOGGER.info("SIMD operations detected as functional. Will replace some operations with faster versions.");
+        } else if (SIMDDetection.versionLimited) {
+            LOGGER.warn("Will not enable SIMD! These optimizations are only safely supported on Java 17+.");
+        } else {
+            LOGGER.warn("SIMD operations are available for your server, but are not configured!");
+            LOGGER.warn("To enable additional optimizations, add \"--add-modules=jdk.incubator.vector\" to your startup flags, BEFORE the \"-jar\".");
+            LOGGER.warn("If you have already added this flag, then SIMD operations are not supported on your JVM or CPU.");
+            LOGGER.warn("Debug: Java: {}, test run: {}", System.getProperty("java.version"), SIMDDetection.testRun);
+        }
+    }
+}
+
