From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kercute <A3167717663@hotmail.com>
Date: Tue, 21 Jan 2025 21:18:25 +0800
Subject: [PATCH] fixup! Blank Configuration File


diff --git a/src/main/java/com/menthamc/mint/config/ConfigInfo.java b/src/main/java/com/menthamc/mint/config/ConfigInfo.java
index e423932cfa17fc74e4911cf094cc70c8735ffe73..8fa8dc7951e2cb475943d2021bd16eb4b3265d9c 100644
--- a/src/main/java/com/menthamc/mint/config/ConfigInfo.java
+++ b/src/main/java/com/menthamc/mint/config/ConfigInfo.java
@@ -8,7 +8,7 @@ import java.lang.annotation.Target;
 @Target(ElementType.FIELD)
 @Retention(RetentionPolicy.RUNTIME)
 public @interface ConfigInfo {
-    String name() default "";
+    String name();
 
     String commit() default "";
 }
