From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kercute <3167717663@qq.com>
Date: Fri, 24 Jan 2025 22:18:03 +0800
Subject: [PATCH] SIMD utilities


diff --git a/build.gradle.kts b/build.gradle.kts
index 50587cde5dcc9cb61ffd8e0a0998a77e2e5c8c3e..1c94c3defd71da5cbb0a23040dae95a255e88a9b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -74,6 +74,13 @@ paperweight {
     craftBukkitPackageVersion.set("v1_21_R1") // also needs to be updated in MappingEnvironment
 }
 
+// Pufferfish Start
+tasks.withType<JavaCompile> {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Pufferfish End
+
 tasks.jar {
     archiveClassifier.set("dev")
 
