From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 22 Mar 2025 02:54:10 +0800
Subject: [PATCH] Add config for vanilla random


diff --git a/src/main/java/dev/bacteriawa/mint/config/modules/fixes/VanillaRandomSourceConfig.java b/src/main/java/dev/bacteriawa/mint/config/modules/fixes/VanillaRandomSourceConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..f10b7fa29a95b0aa79560ba3dcb6a2a91f88a78d
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/config/modules/fixes/VanillaRandomSourceConfig.java
@@ -0,0 +1,11 @@
+package dev.bacteriawa.mint.config.modules.fixes;
+
+import dev.bacteriawa.mint.config.ConfigCategory;
+import dev.bacteriawa.mint.config.ConfigField;
+import dev.bacteriawa.mint.config.Configuration;
+
+@Configuration(name = "use_vanilla_random_source", type = ConfigCategory.fixes)
+public class VanillaRandomSourceConfig {
+    @ConfigField(comment = "Related with RNG cracks")
+    public static boolean enabled = false;
+}
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index feccda50236a3e5f05253d2e8a0742b22bd9fbe4..019c0beaace7100e2bbc7f927919713417fd4f89 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity;
 
+import dev.bacteriawa.mint.config.modules.fixes.VanillaRandomSourceConfig;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableList.Builder;
 import com.google.common.collect.Iterables;
@@ -591,7 +592,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         this.bb = Entity.INITIAL_AABB;
         this.stuckSpeedMultiplier = Vec3.ZERO;
         this.nextStep = 1.0F;
-        this.random = SHARED_RANDOM; // Paper - Share random for entities to make them more random
+        this.random = VanillaRandomSourceConfig.enabled ? RandomSource.create() : SHARED_RANDOM; // Paper - Share random for entities to make them more random // Luminol - Add config for vanilla random SHARED_RANDOM
         this.remainingFireTicks = -this.getFireImmuneTicks();
         this.fluidHeight = new Object2DoubleArrayMap(2);
         this.fluidOnEyes = new HashSet();
