From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sat, 22 Mar 2025 03:20:48 +0800
Subject: [PATCH] Update configuration system and added mint command


diff --git a/src/main/java/dev/bacteriawa/mint/command/MintCommand.java b/src/main/java/dev/bacteriawa/mint/command/MintCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..0ae4bc85f993a34f0bd4b9aa73ed995dbc32e1e0
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/command/MintCommand.java
@@ -0,0 +1,69 @@
+package dev.bacteriawa.mint.command;
+
+import dev.bacteriawa.mint.command.subcommand.ConfigReloadCommand;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.format.NamedTextColor;
+import org.bukkit.command.CommandSender;
+import org.bukkit.command.defaults.BukkitCommand;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MintCommand extends BukkitCommand {
+    private final Map<String, MintSubCommand> subcommands = new HashMap<>();
+
+    public MintCommand() {
+        super("mint");
+        this.setUsage("/commands <usage>");
+
+        subcommands.put("reload", new ConfigReloadCommand());
+    }
+
+    @Override
+    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
+        if (!testPermission(sender) || !sender.isOp()) {
+            sender.sendMessage(Component
+                    .text("No permission to execute this command!")
+                    .color(NamedTextColor.RED)
+            );
+        }
+
+        if (args.length < 1) {
+            sender.sendMessage(Component
+                    .text("Usage: /mint <commands>")
+                    .color(NamedTextColor.RED)
+            );
+        }
+
+        if (args.length >= 1) {
+            String subCommand = args[0];
+
+            if (subcommands.containsKey(subCommand)) {
+                return subcommands.get(subCommand).execute(
+                        sender, subCommand,
+                        Arrays.copyOfRange(args, 1, args.length)
+                );
+            } else {
+                sender.sendMessage(Component
+                        .text("Not exist command: " + subCommand)
+                        .color(NamedTextColor.RED)
+                );
+            }
+        }
+
+        return true;
+    }
+
+    @Override
+    public @NotNull List<String> tabComplete(@NotNull CommandSender sender, @NotNull String alias, @NotNull String[] args) throws IllegalArgumentException {
+        if (args.length == 1){
+            return subcommands.keySet().stream().toList();
+        }
+
+        return List.of();
+    }
+}
+
diff --git a/src/main/java/dev/bacteriawa/mint/command/MintSubCommand.java b/src/main/java/dev/bacteriawa/mint/command/MintSubCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..2a98f65f813fdc6acf745e9dd4424ec859671426
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/command/MintSubCommand.java
@@ -0,0 +1,8 @@
+package dev.bacteriawa.mint.command;
+
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+
+public abstract class MintSubCommand {
+    public abstract boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args);
+}
diff --git a/src/main/java/dev/bacteriawa/mint/command/subcommand/ConfigReloadCommand.java b/src/main/java/dev/bacteriawa/mint/command/subcommand/ConfigReloadCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..449de686b9a5ec7801de131d646c0ccbc1b4a549
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/command/subcommand/ConfigReloadCommand.java
@@ -0,0 +1,14 @@
+package dev.bacteriawa.mint.command.subcommand;
+
+import dev.bacteriawa.mint.command.MintSubCommand;
+import dev.bacteriawa.mint.config.MintConfig;
+import org.bukkit.command.CommandSender;
+import org.jetbrains.annotations.NotNull;
+
+public class ConfigReloadCommand extends MintSubCommand {
+    @Override
+    public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
+        MintConfig.reloadConfig();
+        return true;
+    }
+}
diff --git a/src/main/java/dev/bacteriawa/mint/config/MintConfig.java b/src/main/java/dev/bacteriawa/mint/config/MintConfig.java
index c92c6b78debc9147db46e43addc3f07f5e16a0a1..265d04f51a33f28e03711b18c8215bc170974a0b 100644
--- a/src/main/java/dev/bacteriawa/mint/config/MintConfig.java
+++ b/src/main/java/dev/bacteriawa/mint/config/MintConfig.java
@@ -1,17 +1,20 @@
 package dev.bacteriawa.mint.config;
 
 import com.electronwill.nightconfig.core.file.CommentedFileConfig;
+import dev.bacteriawa.mint.command.MintCommand;
 import io.github.classgraph.ClassGraph;
 import io.github.classgraph.ClassInfo;
 import io.github.classgraph.ScanResult;
+import org.bukkit.Bukkit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.File;
 import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
 import java.nio.charset.StandardCharsets;
-import java.util.Arrays;
 import java.util.Set;
 import java.util.stream.Collectors;
 
@@ -32,6 +35,7 @@ public class MintConfig {
     }
 
     public static void setup() {
+        Bukkit.getCommandMap().register("mint", new MintCommand());
     }
 
     public static void loadConfig() {
@@ -50,44 +54,55 @@ public class MintConfig {
 
     private static void loadConfigInstance(Class<?> moduleClass) {
         try {
-            if (Arrays.stream(moduleClass.getInterfaces())
-                    .collect(Collectors.toSet()).contains(IConfigModule.class)
-            ) {
-                int clazzModifiers = moduleClass.getModifiers();
-                if (!moduleClass.isAnnotationPresent(Configuration.class) || moduleClass.isAnnotationPresent(ConfigSkipLoad.class))
-                    return;
-                if (!(Modifier.isPublic(clazzModifiers) && isPlainClass(moduleClass))) {
-                    LOGGER.error("`{}` must be public and plain class!", moduleClass.getName(), new RuntimeException());
-                    return;
-                }
-
-                Configuration cfg = moduleClass.getDeclaredAnnotation(Configuration.class);
-                for (Field field : moduleClass.getDeclaredFields()) {
-                    int fieldModifiers = field.getModifiers();
-                    if (!field.isAnnotationPresent(ConfigField.class) || field.isAnnotationPresent(ConfigSkipLoad.class))
-                        continue;
-                    if (!(Modifier.isStatic(fieldModifiers) && Modifier.isPublic(fieldModifiers))) {
-                        LOGGER.error("`{}` must be public and static!", field.getName(), new RuntimeException());
-                        continue;
-                    }
+            int clazzModifiers = moduleClass.getModifiers();
+            if (!moduleClass.isAnnotationPresent(Configuration.class) || moduleClass.isAnnotationPresent(ConfigSkipLoad.class))
+                return;
+            if (!(Modifier.isPublic(clazzModifiers) && isPlainClass(moduleClass))) {
+                LOGGER.error("`{}` must be public and plain class!", moduleClass.getName(), new RuntimeException());
+                return;
+            }
 
-                    ConfigField config = field.getDeclaredAnnotation(ConfigField.class);
-                    String fullPath = cfg.type().name() + "." + cfg.name() + "." + field.getName();
+            Configuration cfg = moduleClass.getDeclaredAnnotation(Configuration.class);
+            for (Field field : moduleClass.getDeclaredFields()) {
+                int fieldModifiers = field.getModifiers();
+                if (!field.isAnnotationPresent(ConfigField.class) || field.isAnnotationPresent(ConfigSkipLoad.class))
+                    continue;
+                if (!(Modifier.isStatic(fieldModifiers) && Modifier.isPublic(fieldModifiers))) {
+                    LOGGER.error("`{}` must be public and static!", field.getName(), new RuntimeException());
+                    continue;
+                }
 
-                    if (!configuration.contains(fullPath)) {
-                        configuration.add(fullPath, field.get(null));
+                ConfigField config = field.getDeclaredAnnotation(ConfigField.class);
+                String fullPath = cfg.type().name() + "." + cfg.name() + "." + field.getName();
 
-                        setComment(config.comment(), fullPath);
-                    }
+                if (!configuration.contains(fullPath)) {
+                    configuration.add(fullPath, field.get(null));
 
-                    field.set(null, configuration.get(fullPath));
+                    setComment(config.comment(), fullPath);
                 }
 
-                setComment(cfg.comment(), cfg.type().name() + "." + cfg.name());
+                field.set(null, configuration.get(fullPath));
             }
+
+            setComment(cfg.comment(), cfg.type().name() + "." + cfg.name());
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
+
+        loaded(moduleClass);
+    }
+
+    private static void loaded(Class<?> clazz) {
+        try {
+            Method loadedMethod = clazz.getDeclaredMethod("loaded", CommentedFileConfig.class);
+            int modifiers = loadedMethod.getModifiers();
+            if (!(Modifier.isPublic(modifiers) && Modifier.isStatic(modifiers)))
+                return;
+            loadedMethod.invoke(null, configuration);
+        } catch (NoSuchMethodException ignored) {
+        } catch (InvocationTargetException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
     }
 
     public static void reloadConfig() {
