From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Mon, 10 Jun 2024 10:51:08 +0000
Subject: [PATCH] Try fixing folia off region POI accessing issue


diff --git a/src/main/java/dev/bacteriawa/mint/config/modules/fixes/FoliaPOIAccessOffRegionFixConfig.java b/src/main/java/dev/bacteriawa/mint/config/modules/fixes/FoliaPOIAccessOffRegionFixConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..5ba822e2d05e5c90b61aff4406da2a7bab27e825
--- /dev/null
+++ b/src/main/java/dev/bacteriawa/mint/config/modules/fixes/FoliaPOIAccessOffRegionFixConfig.java
@@ -0,0 +1,12 @@
+package dev.bacteriawa.mint.config.modules.fixes;
+
+import dev.bacteriawa.mint.config.ConfigCategory;
+import dev.bacteriawa.mint.config.ConfigField;
+import dev.bacteriawa.mint.config.Configuration;
+
+@Configuration(name = "folia.fix_poi_access_off_region", type = ConfigCategory.fixes)
+public class FoliaPOIAccessOffRegionFixConfig {
+    @ConfigField(comment = {"Folia's POIManager has some unpatched areas that trigger asynchronous catchers and cause the server to crash.",
+    })
+    public static boolean enabled = true;
+}
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java b/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
index aac1df9e2c0652bc6348af0404deba7465f82d42..4547deeb68f89d42ce2e8da55ef66e9059d8e79e 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/GoToPotentialJobSite.java
@@ -49,10 +49,17 @@ public class GoToPotentialJobSite extends Behavior<Villager> {
             ServerLevel serverLevel2 = serverLevel.getServer().getLevel(pos.dimension());
             if (serverLevel2 != null) {
                 PoiManager poiManager = serverLevel2.getPoiManager();
+                //Luminol start - Fix off region POI accessing
+                java.lang.Runnable scheduledRelease = () -> {
                 if (poiManager.exists(blockPos, poiType -> true)) {
                     poiManager.release(blockPos);
                 }
-
+                };
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverLevel2, blockPos) && dev.bacteriawa.mint.config.modules.fixes.FoliaPOIAccessOffRegionFixConfig.enabled)
+                    io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueTickTaskQueue(serverLevel2, blockPos.getX() >> 4, blockPos.getZ() >> 4, scheduledRelease);
+                else
+                    scheduledRelease.run();
+                //Luminol end
                 DebugPackets.sendPoiTicketCountPacket(serverLevel, blockPos);
             }
         });
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java b/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
index d1a9b62d3304916275dd6b4c4e783cf1563b5e21..b43bab4c653076e9f71eee56ab6bc9485543fffb 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -33,7 +33,10 @@ public class YieldJobSite {
                                 } else if (entity.getVillagerData().getProfession() != VillagerProfession.NONE) {
                                     return false;
                                 } else {
-                                    BlockPos blockPos = context.<GlobalPos>get(potentialJobSite).pos();
+                                    final GlobalPos globalPos = context.<GlobalPos>get(potentialJobSite); //Luminol - Try fixing off main POI accessing
+                                    final net.minecraft.server.level.ServerLevel targetLevel = net.minecraft.server.MinecraftServer.getServer().getLevel(globalPos.dimension()); //Luminol - Try fixing off main POI accessing
+                                    BlockPos blockPos = globalPos.pos(); //Luminol - Try fixing off main POI accessing
+                                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(targetLevel, blockPos) && dev.bacteriawa.mint.config.modules.fixes.FoliaPOIAccessOffRegionFixConfig.enabled) return true; //Luminol - Try fixing off main POI accessing
                                     Optional<Holder<PoiType>> optional = world.getPoiManager().getType(blockPos);
                                     if (optional.isEmpty()) {
                                         return true;
diff --git a/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index f3f98e6276dda3bc4f290fc2d80569f7e1e7ef66..0e59cab6d796a042a236e3dbcdf60bb254e87bb1 100644
--- a/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/src/main/java/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -212,7 +212,7 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
 
     public Stream<PoiRecord> getInSquare(Predicate<Holder<PoiType>> typePredicate, BlockPos pos, int radius, PoiManager.Occupancy occupationStatus) {
         int i = Math.floorDiv(radius, 16) + 1;
-        return ChunkPos.rangeClosed(new ChunkPos(pos), i).flatMap(chunkPos -> this.getInChunk(typePredicate, chunkPos, occupationStatus)).filter(poi -> {
+        return ChunkPos.rangeClosed(new ChunkPos(pos), i).filter(cpos -> dev.bacteriawa.mint.config.modules.fixes.FoliaPOIAccessOffRegionFixConfig.enabled ? ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world,cpos) : true).flatMap(chunkPos -> this.getInChunk(typePredicate, chunkPos, occupationStatus)).filter(poi -> { //Luminol - Fix off region POI access
             BlockPos blockPos2 = poi.getPos();
             return Math.abs(blockPos2.getX() - pos.getX()) <= radius && Math.abs(blockPos2.getZ() - pos.getZ()) <= radius;
         });
